pytorch_lightning/trainer/callback_hook.py

```diff
@@ -257,18 +258,22 @@ def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> Dict[Type, dict]:
             else:
                 state = callback.on_save_checkpoint(self, self.lightning_module, checkpoint)
             if state:
-                callback_states[type(callback)] = state
+                callback_states[callback.state_identifier] = state
         return callback_states
 
-    def on_load_checkpoint(self, checkpoint):
+    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
         """Called when loading a model checkpoint."""
         callback_states = checkpoint.get('callbacks')
+        version = checkpoint.get('pytorch-lightning_version')
         # Todo: the `callback_states` are dropped with TPUSpawn as they
         # can't be saved using `xm.save`
         # https://github.com/pytorch/xla/issues/2773
         if callback_states is not None:
             for callback in self.callbacks:
-                state = callback_states.get(type(callback))
+                state = (
+                    callback_states.get(callback.state_identifier)
+                    or callback_states.get(callback._legacy_state_identifier)
```

Comment:
here is the backward compatibility.

in this PR, I "hardcode" this check here, but I have ideas for a more general way of upgrading checkpoints for backward compatibility. Will discuss in the linked issue.